localstack/services/apigateway/provider_asf.py

```diff
@@ -0,0 +1,185 @@
+from localstack.utils.aws.aws_responses import LambdaResponse
+from localstack.utils.json import parse_json_or_yaml
+from localstack.utils.strings import to_str
+
+
+def to_invocation_context(request: Request) -> ApiInvocationContext:
+    """
+    Converts an HTTP Request object into an ApiInvocationContext.
+
+    :param request: the original request
+    :return: the ApiInvocationContext
+    """
+    # FIXME: ApiInvocationContext should be refactored to use werkzeug request object correctly
+    method = request.method
+    path = request.full_path if request.query_string else request.path
+    # not using request.data here is critical to make sure we do not consume form data
+    data = request.get_data(cache=True) or b""
+    headers = Headers(request.headers)
+
+    # TODO: here will have to be a compatibility layer for the ASF gateway
+
+    return ApiInvocationContext(
+        method,
+        path,
+        data,
+        headers,
+    )
+
+
+class ApigatewayRouter:
+    """
+    Simple implementation around a Router to manage dynamic restapi routes (routes added by a user through the
+    apigateway API).
+    """
+
+    router: Router[Handler]
+
+    def __init__(self, router: Router[Handler]):
+        self.op_router = RestServiceOperationRouter(load_service("apigateway"))
+        self.router_rules: Dict[str, List[Rule]] = defaultdict(list)
+        self.router = router
+
+    def add_rest_api(self, rest_api: RestApi) -> None:
+        """
+        Adds a route for the given RestApi.
+        :param rest_api: the RestApi to add
+        """
+        # TODO: probably it is better to have a parameterized handler and only one rule, rather than creating a new
+        #  rule for every rest API. the more rules there are, the slower the routing will be. on the other hand,
+        #  regex matching could be just as slow. need to check!
+        api_id = rest_api["id"]
+
+        # add the canonical execute-api handler
+        self.router_rules[api_id].append(
+            self.router.add(
+                "/",
+                host=f"{api_id}.execute-api.<regex('.*'):server>",
+                endpoint=self._restapis_host_handler,
+            )
+        )
+        self.router_rules[api_id].append(
+            self.router.add(
+                "/<path:path>",
+                host=f"{api_id}.execute-api.<regex('.*'):server>",
+                endpoint=self._restapis_host_handler,
+            )
+        )
+
+        # add the localstack-specific _user_request_ handler
+        self.router_rules[api_id].append(
+            self.router.add(
+                f"/restapis/{api_id}/<stage>/_user_request_",
+                endpoint=self._restapis_user_request_handler,
+            )
+        )
+        self.router_rules[api_id].append(
+            self.router.add(
+                f"/restapis/{api_id}/<stage>/_user_request_/<path:path>",
+                endpoint=self._restapis_user_request_handler,
+            )
+        )
```

Comment:
nit: Since we are now registering the routes using the router, we should add a deprecation comment to the section in the service router which currently deals with these routes:
https://github.com/localstack/localstack/blob/de2bf6652320edcadcf461057a6cdf7ee3cc9fb0/localstack/aws/protocol/service_router.py#L153-L157

Reply:
added a TODO here (don't know where the commit you linked to is though)